tutorials/sphinx-tutorials/multiagent_competitive_ddpg.py

```diff
@@ -0,0 +1,992 @@
+# with the only difference that it will auto reset done states until the desired frames are collected.
+#
+# We need to feed it our exploration policies. Furthermore, to run the policies from all groups as if they were one,
+# we put them in a sequence. They will not interfere with each other as each group writes and reads keys in different places.
+#
+
+# Put exploration policies from each group in a sequence
+agents_exploration_policy = TensorDictSequential(*exploration_policies.values())
+
+collector = SyncDataCollector(
+    env,
+    agents_exploration_policy,
+    device=device,
+    frames_per_batch=frames_per_batch,
+    total_frames=total_frames,
+)
+
+######################################################################
+# Replay buffer
+# -------------
+#
+# Replay buffers are a common building piece of off-policy RL sota-implementations.
+# There are many types of buffers, in this tutorial we use a basic buffer to store and sample tensordict
+# data randomly.
+#
+
+replay_buffers = {}
+for group, _agents in env.group_map.items():
+    replay_buffer = ReplayBuffer(
+        storage=LazyTensorStorage(
+            memory_size, device=device
+        ),  # We will store up to memory_size multi-agent transitions
+        sampler=RandomSampler(),
+        batch_size=train_batch_size,  # We will sample batches of this size
+    )
+    replay_buffers[group] = replay_buffer
+
+######################################################################
+# Loss function
+# -------------
+#
+# The DDPG loss can be directly imported from TorchRL for convenience using the
+# :class:`~.objectives.DDPGLoss` class. This is the easiest way of utilising DDPG:
+# it hides away the mathematical operations of DDPG and the control flow that
+# goes with it.
+#
+# It is also possible to have different policies for each group.
+#
+losses = {}
+for group, _agents in env.group_map.items():
+    loss_module = DDPGLoss(
+        actor_network=policies[group],  # Use the non-explorative policies
+        value_network=critics[group],
+        delay_value=True,  # Whether to use a target network for the value
+        loss_function="l2",
+    )
+    loss_module.set_keys(
+        state_action_value=(group, "state_action_value"),
+        reward=(group, "reward"),
+        done=(group, "done"),
+        terminated=(group, "terminated"),
+    )
+    loss_module.make_value_estimator(ValueEstimators.TD0, gamma=gamma)
+
+    losses[group] = loss_module
+
+target_updaters = {
+    group: SoftUpdate(loss, tau=polyak_tau) for group, loss in losses.items()
+}
+
+optimizers = {
+    group: {
+        "loss_actor": torch.optim.Adam(
+            loss.actor_network_params.flatten_keys().values(), lr=lr
+        ),
+        "loss_value": torch.optim.Adam(
+            loss.value_network_params.flatten_keys().values(), lr=lr
+        ),
+    }
+    for group, loss in losses.items()
+}
```

Comment:
when I see that we make a different loss, a different rb and a different optimizer for each group I wonder if we aren't doing things wrong. This looks like smth that could be simplified. 
After reading this code and knowing the library, I'm pretty sure I couldn't get the whole script right the first time!

Reply:
In other words, I'd like to understand why we can't just have one RB, one loss and one optimizer.

Reply:
Plus we want to allow different groups to have different buffers losses and lrs